core/network/impl/rpc_sender_libp2p.cpp

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "network/impl/rpc_sender_libp2p.hpp"
+
+#include "libp2p/basic/message_read_writer.hpp"
+
+namespace kagome::network {
+  using libp2p::basic::MessageReadWriter;
+
+  RPCSenderLibp2p::RPCSenderLibp2p(libp2p::Host &host, common::Logger logger)
+      : host_{host}, log_{std::move(logger)} {}
+
+  void RPCSenderLibp2p::sendWithResponse(RPCInfoLibp2p rpc_info,
+                                         BufferSPtr request,
+                                         Callback cb) const {
+    host_.newStream(
+        rpc_info.peer_info,
+        rpc_info.protocol,
+        [self{shared_from_this()},
+         request{std::move(request)},
+         cb{std::move(cb)}](auto &&stream_res) mutable {
+          if (!stream_res) {
+            self->log_->error("cannot open a stream");
+            return cb(stream_res.error());
+          }
+
+          auto read_writer = std::make_shared<MessageReadWriter>(
+              std::move(stream_res.value()));
+          read_writer->write(
+              *request,
+              [self{std::move(self)}, request, read_writer, cb{std::move(cb)}](
+                  auto write_res) mutable {
+                self->receive(std::move(write_res), read_writer, std::move(cb));
```

Comment:
Why don't you handle error here instead of passing it to receive, moreover adding excessive parameter?